### PR TITLE
Speedup stdcm editoast endpoint :rocket: 

### LINKED ISSUE #10452: STDCM queries take too much time
### What happened?

See the traces of an stdcm query:

Image

Private datalog link to [the trace](https://claps.datadoghq.eu/apm/traces?query=service%3Aosrd-editoast%20%40deployment.environment.name%3Aosd-prod%20stdcm&agg_m=count&agg_m_source=base&agg_t=count&cols=core_service%2Ccore_resource_name%2Clog_duration%2Clog_http.method%2Clog_http.status_code&fromUser=false&graphType=waterfall&historicalData=false&messageDisplay=inline&panel_tab=flamegraph&query_translation_version=v0&shouldShowLegend=true&sort=desc&spanID=1280736463641084366&spanType=all&spanViewType=metadata&storage=hot&timeHint=1737391622862&trace=286e95605c7341863da96998498b7f4510254662739427355806&traceID=286e95605c7341863da96998498b7f45&view=spans&start=1737391780016&end=1737392680016&paused=false).

- [ ] Add `tracing::instrument` for critical functions.
- [ ] Compress the data to valkey
- [ ] Optimize the psql train schedule retrieve query
- [ ] Optimize the `PathItemCache::load` function (unnecessary queries when we hit cache 100%)
    - Skip query if no items
    - Use HashSet instead to avoid duplicates 
- [ ] Valkey queries seem to be executed sequentially. We must fix it so each batch can run in parallel.
    - `Tokio(ConnectionManager)` can handle only one physical connection.
    - We could use [deadpool-redis](https://crates.io/crates/deadpool-redis)

### What did you expect to happen?

Retrieving the simulation from valkey should be way more efficient.

- Check that the filter is working properly
- Check the size of the payload (is it transfer time)

Storing the stdcm log takes too much time:
- We are supposed to spawn a worker to insert in DB the input and output of the query.

### How can we reproduce it (as minimally and precisely as possible)?

1. Do a stdcm with a real timetable and infra
2. Observe traces

### On which environments the bug occurs?

Production (SNCF)

### On which browser the bug occurs?

Firefox

### OSRD version (top right corner `Account` button > `Informations`)

c9066aa

### PR DESCRIPTION
**Pull request by flomonster**
close #10452

What si done (review commits individual):

1. Spawn tasks to dispatch cpu charge
2. Reduce train batch size
3. Avoid useless psql queries
4. Optimize a postgresql queries. Limit the number of elements in the response. Moving a filter from rust to postgres. (check perf here: https://explain.dalibo.com/plan/e7d4427b7a8ea7d5)

### Local benchmark

|               | Before | After |
|---------------|--------|-------|
| Without cache |  106s  |  108s |
| With cache    |  9.64s | 3.15s |

### Dev benchmark

|               | Before (Compression) | After (No-Compression) | After (Compression) |
|---------------|----------------------|------------------------|---------------------|
| Without cache | 90s                  |          111s          |         109s        |
| With cache    | 17s                  |          12.7s         |         8.3s        |

### CODE REVIEW

File: editoast/src/views/train_schedule.rs [hunk 1]
```diff
@@ -404,30 +405,37 @@ pub async fn train_simulation_batch(
...
-                    core.clone(),
-                    infra,
-                    chunk,
-                    consists_ref,
-                    electrical_profile_set_id,
-                )
-                .await
-            }
+            let consists = consists.clone();
+            let infra = <Infra as Clone>::clone(infra);
```
Comment (woshilapin) [inline editoast/src/views/train_schedule.rs:414]: `infra.clone()` doesn't work? 🤔

File: editoast/src/models/timetable.rs [hunk 2]
```diff
@@ -55,27 +56,64 @@ impl Timetable {
...
+    ///
+    /// ```
+    ///                       min_time               max_time
+    ///     Time Window           |----------------------|
+    /// Train 1 ✅         |-------------|
+    /// Train 2 ❌    |-------|
+    /// Train 3 ✅                              |----------|
+    /// Train 4 ✅   |-------?
+    /// Train 5 ✅                        |---------?
+    /// Train 6 ❌                                           |------?
```
Comment (woshilapin) [inline editoast/src/models/timetable.rs:75]: Nice ASCII art. Since there is a lot edge case we want to support, this looks like it'd be nice to make some tests for all these edge cases.

File: editoast/src/views/timetable/stdcm.rs [hunk 3]
```diff
@@ -216,65 +216,10 @@ async fn stdcm(
...
-    // For trains departing before the LMR train departure (tA):
-
-    // If the train's arrival time is unscheduled (?), the train is kept (e.g., Train 4 and Train 5).
-    // If the train's arrival time is scheduled (|), the train is kept only if its arrival time is after the LMR train's earliest departure time.
-    // Train 1 is kept and train 2 is filtered out.
-
-    // Step 1
-    let mut train_schedules = timetable
-        .schedules_before_date(&mut conn, latest_simulation_end)
+    let train_schedules = timetable
```
Comment (woshilapin) [inline editoast/src/views/timetable/stdcm.rs:219]: `// Step 1` has been removed as well as `// Step 2 and 3` below. Other steps (comments) of the function might need to be updated.

File: editoast/src/models/timetable.rs [hunk 4]
```diff
@@ -55,27 +56,64 @@ impl Timetable {
...
-            .await?
-            .map_ok(|ts| ts.into())
-            .try_collect::<Vec<TrainSchedule>>()
-            .await;
-        match train_schedules {
-            Ok(train_schedules) => Ok(train_schedules),
-            Err(err) => Err(err.into()),
-        }
+        let train_schedules = sql_query("
+        WITH last_path_step AS (
```
Comment (woshilapin) [inline editoast/src/models/timetable.rs]: For big SQL queries, we try to put them in external files (helps also with syntax highlighting in our editors). See [here](https://github.com/OpenRailAssociation/osrd/blob/14fa21bf67cf757bd42881dbf5459adab5281a8c/editoast/src/models/infra/voltage.rs#L27) for example.

**Review by woshilapin — APPROVED** [review]
Nice improvement of performance 🚀 Only minor comments.

File: editoast/src/views/train_schedule.rs (hunk 1 above)
Comment (flomonster) [reply editoast/src/views/train_schedule.rs:414]: Nope `Infra::clone` exists

File: editoast/src/views/timetable/stdcm.rs (hunk 3 above)
Comment (flomonster) [reply editoast/src/views/timetable/stdcm.rs:219]: These steps were only for filtering the timetable. No need to update other comments.

File: editoast/src/views/train_schedule.rs (hunk 1 above)
Comment (woshilapin) [reply editoast/src/views/train_schedule.rs:414]: What an unfortunate naming of the function 😄

File: editoast/src/models/timetable.rs (hunk 4 above)
Comment (flomonster) [reply editoast/src/models/timetable.rs]: Done

File: editoast/src/views/train_schedule.rs (hunk 1 above)
Comment (flomonster) [reply editoast/src/views/train_schedule.rs:414]: Yeah, it should probably be renamed `Infra::duplicate` or something like that.

File: editoast/src/models/timetable.rs (hunk 2 above)
Comment (flomonster) [reply editoast/src/models/timetable.rs:75]: Test added
